File: src/cn/langya/sun/modules/impl/combat/KillAura.java
package cn.langya.sun.modules.impl.combat;

import cn.langya.sun.events.impl.player.EventUpdate;
import cn.langya.sun.modules.Category;
import cn.langya.sun.modules.Module;
import cn.langya.sun.utils.misc.MathUtil;
import cn.langya.sun.utils.misc.TimeUtil;
import cn.langya.sun.utils.player.RotationUtil;
import cn.langya.sun.values.BoolValue;
import cn.langya.sun.values.DoubleValue;
import cn.langya.sun.values.IntValue;
import cn.langya.sun.values.ListValue;
import com.cubk.event.annotations.EventTarget;
import de.florianmichael.viamcp.fixes.AttackOrder;
import lombok.Getter;
import lombok.Setter;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.monster.EntityMob;
import net.minecraft.entity.player.EntityPlayer;

import java.util.ArrayList;
import java.util.List;

/**
 * @author LangYa
 * @date 2024/2/4 ÏÂÎç 07:31
 */

@Getter
public class KillAura extends Module {

    private static final IntValue minCpsValue = new IntValue("MinCPS", 10, 20, 1);
    private static final IntValue maxCpsValue = new IntValue("MaxCPS", 10, 20, minCpsValue.get());
    private static final DoubleValue groundRangeValue = new DoubleValue("GroudRange", 3.0, 6, 3.0);
    private static final DoubleValue airRangeValue = new DoubleValue("AirRange", 3.0, 6, 3.0);
    private static final ListValue targetModeValue = new ListValue("TargetMode", "Single");
    private static final BoolValue rotationValue = new BoolValue("Rotation", true);
    private static final ListValue rotationModeValue = new ListValue("RotationsMode", "Vanilla");
    private static final BoolValue rayCastValue = new BoolValue("RayCast", true);
    private static final BoolValue swingValue = new BoolValue("Swing", true);
    private static final BoolValue keepSprintValue = new BoolValue("KeepSprint", false);
    private static final BoolValue targetPlayerValue = new BoolValue("TargetPlayer", true);
    private static final BoolValue targetMobValue = new BoolValue("TargetMob", false);

    public KillAura() {
        super("KillAura", Category.Combat);
        add(minCpsValue,maxCpsValue,groundRangeValue, airRangeValue, targetModeValue, rotationValue, rotationModeValue, rayCastValue, swingValue, keepSprintValue);

        targetModeValue.getValues().add("Single");
        targetModeValue.getValues().add("Multi");

        rotationModeValue.getValues().add("Vanilla");
        rotationModeValue.getValues().add("GrimAC");

        targets = new ArrayList<>();

    }

    @Getter
    private static List<EntityLivingBase> targets;
    @Setter
    private static EntityLivingBase target;
    @Setter
    // yaw pitch
    private static float[] rotations = new float[]{0, 0};
    @Getter
    private static TimeUtil attacktimer = new TimeUtil();

    @EventTarget
    void onU(EventUpdate e) {
        targets.clear();

        for (Entity entity : mc.world.loadedEntityList) {
            if (entity instanceof EntityLivingBase && mc.player.getDistanceToEntity(entity) <= getRange() && !entity.isDead && entity != mc.player) {
                targets.add((EntityLivingBase) entity);
            }
        }

        //targets.stream().filter(entity -> mc.player.getDistanceToEntity(entity) >= getRange() || entity.getHealth() > 0 || entity != mc.player);

        if(!targetPlayerValue.get()) {
            targets.removeIf(entity -> entity instanceof EntityPlayer);
        }

        if(!targetMobValue.get()) {
            targets.removeIf(entity -> entity instanceof EntityMob);
        }

        if(targets.isEmpty()) return;

        switch (targetModeValue.get()) {
            case "Single":
                target = targets.get(0);
                break;
            case "Multi":
                for (EntityLivingBase entity : targets) {
                    target = entity; // ?
                }
                break;
        }


        if (rayCastValue.get() && rotationValue.get() && !RotationUtil.isMouseOver(e.getYaw(), e.pitch, target, getRange()))
            return;


        if (rotationValue.get()) e.setRotations(getRotations());

        if(attacktimer.hasTimePassed(1000 / getCps())) {
            attackEntity(target);
            attacktimer.reset();
        }


    }

    static void attackEntity(EntityLivingBase e) {

        if (targetModeValue.get().equals("Single") && mc.player.getAttackingEntity() != null) {
            return;
        }

        AttackOrder.sendFixedAttack(mc.player, e);

    }

    static float[] getRotations() {
        if (rotationValue.get()) {
            switch (rotationModeValue.get()) {
                case "Vanilla":
                    rotations = RotationUtil.getTargetRotations(target);
                    break;
                case "GrimAC":
                    rotations = RotationUtil.getGrimRotations(target);
                    break;
            }
        }

        return rotations;
    }

    static double getRange() {
        return mc.player.onGround ? groundRangeValue.get() : airRangeValue.get();
    }

    static int getCps() {
        final int maxValue = (maxCpsValue.getMaximum() - maxCpsValue.get()) * 20;
        final int minValue = (minCpsValue.getMaximum() - minCpsValue.get()) * 20;
        return MathUtil.getRandomInRange(minValue, maxValue);
    }

}

File: src/cn/langya/sun/modules/impl/combat/KillAura.kt
package cn.langya.sun.modules.impl.combat

import cn.langya.sun.events.impl.Render3DEvent
import cn.langya.sun.events.impl.UpdateEvent
import cn.langya.sun.modules.Category
import cn.langya.sun.modules.Module
import cn.langya.sun.modules.impl.misc.Teams
import cn.langya.sun.utils.misc.TimeUtil
import cn.langya.sun.utils.player.RotationUtil
import cn.langya.sun.values.BoolValue
import cn.langya.sun.values.FloatValue
import cn.langya.sun.values.IntValue
import com.cubk.event.annotations.EventTarget
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.network.play.client.CPacketPlayerDigging
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import org.lwjgl.opengl.GL11
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin


class KillAura : Module("KillAura",Category.Combat) {
    //最大攻击速度
    private val maxCPSValue = IntValue("MaxCPS", 8, 20, 0)

    //最小攻击速度
    private val minCPSValue = IntValue("MinCPS", 5, 19, 0)

    //攻击距离
    val groundrangeValue = FloatValue("GroundRange", 3F, 8F, 3F)
    val airrangeValue = FloatValue("AirRange", 3F, 8F, 3F)

    //攻击是否穿墙
    private val throughWalls = BoolValue("ThroughWalls", true)

    //挥手
    private val swingValue = BoolValue("Swing", true)

    //AB
    private val autoBlockvalue = BoolValue("AutoBlock", true)

    //静默转头
    private val silentrotationValue = BoolValue("SilentRotation", true)

    //最大转头速度
    private val maxTurnSpeed = FloatValue("MaxTurnSpeed", 180f, 180f, 0f)

    //最小转头速度
    private val minTurnSpeed = FloatValue("MinTurnSpeed", 180f, 180f, 0f)

    //移动修复
    private val movefixValue = BoolValue("MoveFix", true)

    //视角
    private val fovValue = FloatValue("FOV", 180f, 180f, 0f)

    //距离光环显示
    private val circleValue = BoolValue("Circle", true)
    private val circleRed = IntValue("CircleRed", 255, 255, 0)
    private val circleGreen = IntValue("CircleGreen", 255, 255, 0)
    private val circleBlue = IntValue("CircleBlue", 255, 255, 0)
    private val circleAlpha = IntValue("CircleAlpha", 255, 255, 0)
    private val circleAccuracy = IntValue("CircleAccuracy", 15, 60, 0)

    //打人光环显示
    private val mark = BoolValue("Mark", true)
    init {
        add(maxCPSValue,minCPSValue,groundrangeValue,airrangeValue,throughWalls,swingValue,autoBlockvalue,silentrotationValue,maxTurnSpeed,minTurnSpeed,movefixValue,fovValue,circleValue,circleRed,circleGreen,circleBlue,circleAlpha,circleAccuracy,mark)
    }

    val attackTimer = TimeUtil()

    public var target: Entity? = null
    var click = 0
    var blocking = false


    @EventTarget
    fun onUpdate(event: UpdateEvent) {

        if(!state || mc.player == null || mc.world == null) return

        for(entity in mc.world.loadedEntityList) {
            if (mc.player.getDistanceToEntity(entity) <= getRange() && !Teams.isSameTeam(entity) && entity != mc.player &&  attackTimer.hasTimePassed(randomClickDelay(minCPSValue.get(), maxCPSValue.get())) && !Teams.isSameTeam(entity) && isFovInRange(entity) && !entity.isDead) {
                target = entity
                attackTimer.reset()
                attackEntity(entity)
            } else {
                target = null
            }
        }


        if (target == null) {
            stopBlocking()
            click = 0
        }

        if(mc.player.attackingEntity == null) {
            startBlock()
        }

    }

    @EventTarget
    fun onRender3D(event: Render3DEvent) {
        if(!state) return

        if (circleValue.get()) {
            GL11.glPushMatrix()
            GL11.glTranslated(
                mc.player.lastTickPosX + (mc.player.posX - mc.player.lastTickPosX) * mc.renderPartialTicks - mc.renderManager.renderPosX,
                mc.player.lastTickPosY + (mc.player.posY - mc.player.lastTickPosY) * mc.renderPartialTicks - mc.renderManager.renderPosY,
                mc.player.lastTickPosZ + (mc.player.posZ - mc.player.lastTickPosZ) * mc.renderPartialTicks - mc.renderManager.renderPosZ
            )
            GL11.glEnable(GL11.GL_BLEND)
            GL11.glEnable(GL11.GL_LINE_SMOOTH)
            GL11.glDisable(GL11.GL_TEXTURE_2D)
            GL11.glDisable(GL11.GL_DEPTH_TEST)
            GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA)

            GL11.glLineWidth(1F)
            GL11.glColor4f(
                circleRed.get().toFloat() / 255.0F,
                circleGreen.get().toFloat() / 255.0F,
                circleBlue.get().toFloat() / 255.0F,
                circleAlpha.get().toFloat() / 255.0F
            )
            GL11.glRotatef(90F, 1F, 0F, 0F)
            GL11.glBegin(GL11.GL_LINE_STRIP)

            for (i in 0..360 step 61 - circleAccuracy.get()) { // You can change circle accuracy  (60 - accuracy)
                GL11.glVertex2f(
                    (cos(i * Math.PI / 180.0).toFloat() * getRange()),
                    ((sin(i * Math.PI / 180.0).toFloat() * getRange()))
                )
            }
            GL11.glVertex2f(
                (cos(360 * Math.PI / 180.0).toFloat() * getRange()).toFloat(),
                ((sin(360 * Math.PI / 180.0).toFloat() * getRange()).toFloat())
            )

            GL11.glEnd()

            GL11.glDisable(GL11.GL_BLEND)
            GL11.glEnable(GL11.GL_TEXTURE_2D)
            GL11.glEnable(GL11.GL_DEPTH_TEST)
            GL11.glDisable(GL11.GL_LINE_SMOOTH)

            GL11.glPopMatrix()
        }

    }

    public fun getRange(): Float {
        return if(mc.player.onGround) {
            groundrangeValue.get()
        } else {
            airrangeValue.get()
        }
    }

    // CPS
    private fun randomClickDelay(minCPS: Int, maxCPS: Int): Long {
        return (Math.random() * (1000 / minCPS - 1000 / maxCPS + 1) + 1000 / maxCPS).toLong()
    }

    //视角
    private fun isFovInRange(entity: Entity?, fov: Float = fovValue.get()): Boolean {

        var fov = fov
        fov *= 0.5.toFloat()
        val v: Double =
            ((mc.player.rotationYaw - getPlayerRotation(entity)) % 360.0 + 540.0) % 360.0 - 180.0
        return v > 0.0 && v < fov || -fov < v && v < 0.0
    }

    //获取玩家转头
    private fun getPlayerRotation(entity: Entity?): Float {
        val x: Double = entity!!.posX - mc.player.posX
        val z: Double = entity.posZ - mc.player.posZ
        var yaw = atan2(x, z) * 57.2957795
        yaw = -yaw
        return yaw.toFloat()
    }

    private fun startBlock() {

        if(!autoBlockvalue.get()) {
            return
        }


        mc.connection!!.networkManager.sendPacket(CPacketPlayerDigging(
            CPacketPlayerDigging.Action.START_DESTROY_BLOCK,
            BlockPos(mc.player.posX, mc.player.posY, mc.player.posZ),
            EnumFacing.UP))

    }


    //  攻击
    private fun attackEntity(entity: Entity?, swing: Boolean = swingValue.get()) {

        if (mc.player.attackingEntity != null) {
            return
        }

        if(!isFovInRange(entity)) {
            return
        }

        doRotation(target as EntityLivingBase)

        stopBlocking()

        if(swing) {
            mc.player.swingArm(EnumHand.MAIN_HAND)
        }

        mc.playerController.attackEntity(mc.player, entity!!)

    }

    private fun doRotation(target: EntityLivingBase) {
        RotationUtil.setVisualRotations(mc.player.rotationYaw)
        val rotations: FloatArray = RotationUtil.getGrimRotations(target)
        mc.player.rotationYaw = rotations[0]
        mc.player.rotationPitch = rotations[1]
    }

    private fun stopBlocking() {

        if (mc.player.isUsingItem || blocking) {
            mc.connection!!.sendPacket(
                CPacketPlayerDigging(
                    CPacketPlayerDigging.Action.RELEASE_USE_ITEM,
                    BlockPos.ORIGIN, EnumFacing.DOWN
                )
            )
            blocking = false
        }

    }

}
